scipy/spatial/_geometric_slerp.py

```diff
@@ -0,0 +1,206 @@
+    ...                     opposite_pole,
+    ...                     t_vals)
+    ... except ValueError:
+    ...     print("ValueError")
+    ValueError
+
+
+    Extend the original example to a sphere and plot interpolation
+    points in 3D:
+
+    >>> from mpl_toolkits.mplot3d import proj3d
+    >>> fig = plt.figure()
+    >>> ax = fig.add_subplot(111, projection='3d')
+
+    Plot the unit sphere for reference (optional):
+
+    >>> u = np.linspace(0, 2 * np.pi, 100)
+    >>> v = np.linspace(0, np.pi, 100)
+    >>> x = np.outer(np.cos(u), np.sin(v))
+    >>> y = np.outer(np.sin(u), np.sin(v))
+    >>> z = np.outer(np.ones(np.size(u)), np.cos(v))
+    >>> ax.plot_surface(x, y, z, color='y', alpha=0.1)
+
+    Interpolating over a larger number of points
+    may provide the appearance of a smooth curve on
+    the surface of the sphere, which is also useful
+    for discretized integration calculations on a
+    sphere surface:
+
+    >>> start = np.array([1, 0, 0])
+    >>> end = np.array([0, 0, 1])
+    >>> t_vals = np.linspace(0, 1, 200)
+    >>> result = geometric_slerp(start,
+    ...                          end,
+    ...                          t_vals)
+    >>> ax.plot(result[...,0],
+    ...         result[...,1],
+    ...         result[...,2],
+    ...         c='k')
+    >>> plt.show()
+    """
+
+    start = np.asanyarray(start)
+    end = np.asanyarray(end)
+
+    if start.ndim != 1 or end.ndim != 1:
+        raise ValueError("Start and end coordinates must be flat")
+
+    if start.size != end.size:
+        raise ValueError("The dimensions of start and "
+                         "end must match (have same size)")
+
+    if start.size == 0 or end.size == 0:
+        raise ValueError("One or both input coordinates are empty")
+
+    if np.array_equal(start, end):
+        raise ValueError("Start and end coordinates cannot be the same")
+
+    if start.size == 1 or end.size == 1:
+        raise ValueError("Cannot interpolate the 0-sphere")
+
+    if not isinstance(tol, float):
+        raise ValueError("tol must be a float")
+    else:
+        tol = math.fabs(tol)
+
+    coord_dist = euclidean(start, end)
+
+    # diameter of 2 within tolerance means antipodes, which is a problem
+    # for all unit n-spheres (even the 0-sphere would have an ambiguous path)
+    if np.allclose(coord_dist, 2.0, rtol=0, atol=tol):
+        raise ValueError("start and end are antipodes"
+                         " using the specified tolerance or they"
+                         " are not on a unit n-sphere")
+
+    # a diameter > 2 within tol violates requirement for input to be on
+    # unit n-sphere
+    if coord_dist > (2.0 + tol):
+        raise ValueError("start and end are not"
+                         " on a unit n-sphere")
+
+    # similarly for points that violate equation for n-sphere
+    # NOTE: this is tricky to think about for the 0-sphere;
+    # performing the check for 1-sphere (circle) and up
+    if start.size > 1:
```

Comment:
From my perspective this check must go first. And I suggest that you should use `np.linalg.norm` for simplicity.

As an alternative, we can do the normalization here instead of checking.

Reply:
Why should this check go first? The transplant causes 7 tests failures, and I'm not clear that this violation is definitely more pathological than some of the others that precede it. 

Unless we have useful information on the average input this function is likely to receive over time, it seems to be somewhat speculative to reorder the error handling.

I think I'd prefer to avoid normalization & have the function focus on being good at a narrow range of things.

Reply:
I meant before the two previous checks (involving `tol`). To me it looks natural to check that `start` and `end` belongs to a unit sphere first.

> I think I'd prefer to avoid normalization & have the function focus on being good at a narrow range of things.

I think there is no single best solution here --- checking and normalization are two viable options. I'd prefer normalization with an explanation in the docstring. There is also an option to provide a flag like `normalize=False` --- either a user guarantees that he normalized the input or asks us to do that.

Reply:
I did the `np.linalg.norm` thing but I may still need convincing that it is worth reordering the exception handling.

Reply:
I reordered the exception handling as requested here, catching unit sphere violations before the two tolerance-related checks---so marking as resolved.